src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs

```diff
@@ -317,44 +317,58 @@ public async Task DuplicateAndClose_TcpServerHandler(AddressFamily addressFamily
-                await using NamedPipeServerStream pipeServerStream =
-                    new NamedPipeServerStream(_ipcPipeName, PipeDirection.Out);
+                // socket used to exchange duplicated socket info with server code
+                using Socket ipcServerListener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                ipcServerListener.BindToAnonymousPort(IPAddress.Loopback);
+                ipcServerListener.Listen(1);
+                string ipcPortString = ((IPEndPoint)ipcServerListener.LocalEndPoint).Port.ToString(CultureInfo.InvariantCulture);
 
                 if (sameProcess)
                 {
-                    Task handlerCode = Task.Run(() => HandlerServerCode(_ipcPipeName));
-                    await RunCommonHostLogic(Environment.ProcessId);
+                    Task handlerCode = Task.Run(() => HandlerServerCode(ipcPortString));
+                    RunCommonHostLogic(Environment.ProcessId);
                     await handlerCode;
                 }
                 else
                 {
-                    RemoteInvokeHandle hServerProc = RemoteExecutor.Invoke(HandlerServerCode, _ipcPipeName);
-                    await RunCommonHostLogic(hServerProc.Process.Id);
-                    await hServerProc.DisposeAsync();
+                    RemoteInvokeHandle hServerProc = RemoteExecutor.Invoke(HandlerServerCode, ipcPortString);
+
+                    // Since RunCommonHostLogic can throw, we need to make sure the server process is disposed
+                    try
+                    {
+                        RunCommonHostLogic(hServerProc.Process.Id);
+                    }
+                    finally
+                    {
+                        await hServerProc.DisposeAsync();
+                    }
                 }
 
-                async Task RunCommonHostLogic(int processId)
+                void RunCommonHostLogic(int processId)
                 {
-                    await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
-                    
-                    pipeServerStream.WaitForConnection();
+                    using Socket ipcServer = ipcServerListener.Accept();
+                    ipcServer.NoDelay = true;
 
                     // Duplicate the socket:
                     SocketInformation socketInfo = handlerOriginal.DuplicateAndClose(processId);
-                    WriteSocketInfo(pipeServerStream, socketInfo);
+                    // Send socketInfo to server code
+                    using var ipcStream = new NetworkStream(ipcServer, false);
+                    WriteSocketInfo(new NetworkStream(ipcServer), socketInfo);
 
                     // Send client data:
                     client.Send(TestBytes);
                 }
 
-                static async Task<int> HandlerServerCode(string ipcPipeName)
+                static async Task<int> HandlerServerCode(string ipcPortString)
```

Comment:
Just a question, is there a reason why we don't pass serialized `SocketInfo` directly here as a string? Completely avoiding the extra socket?
Is it change in timing of the test?

Reply:
Yes, I was hoping for that too, but then realized that `DuplicateAndClose` requires process ID of the other, in our case remote, process to return SocketInfo. So the other process has to start before we obtain socket info, and we hence can't pass serialized socket into to it during its construction.